contributors/design-proposals/containerized-mounter-pod.md

```diff
@@ -0,0 +1,109 @@
+
+To sum it up, it's just a daemon set that spawns privileged pods with some labels, running a simple init and waiting for Kubernetes to do `kubectl exec <the pod> <some utility> <args>`.
+
+## Design
+
+### Configuration of the host OS
+With the [HostPath volume propagation](propagation.md) implemented, we must ensure that `/var/lib/kubelet` is share-able into containers so a pod with mount utilities can mount something there in its container and this mount will be visible to kubelet and docker (or other container engine) on the host.
+We propose:
+*  During startup. kubelet checks that `/var/lib/kubelet` on the host is on a mount with shared mount propagation.
+    *  If not, kubelet makes `/var/lib/kubelet` shareable:
+       ```shell
+       mount --bind /var/lib/kubelet /var/lib/kubelet
+       mount --make-shared /var/lib/kubelet
+       ```
+       Note that these commands are executed on the host in case kubelet runs in a container.
+       If this startup adjustment fails, kubelet refuses to start with a clear error message.
+* Kubelet does not do any other special check. We assume that kubelet either runs on the host and in the same mount namespace as docker and the host, or it runs in a container that sees /var/lib/kubelet as it is on the host and with shared mount propagation.
+
+This ensures that kubelet runs out of the box on any distro without any configuration done by the cluster admin.
+
+### Volume plugins
+* All volume plugins need to be updated to use a new `VolumeExec` interface to call external utilities like `mount`, `mkfs`, `rbd lock` and such. Implementation of the interface will be provided by caller and will lead either to `exec` on the host or `kubectl exec` or `docker exec` in a remote or local pod with utilities for appropriate volume plugin (or docker-exec-like command if another container engine is used).
+
+### Controller
+* There will be new parameter to kube-controller-manager and kubelet:
+    * `--experimental-mount-namespace`, which specifies a dedicated namespace where all pods with mount utilities reside.
+    * `--experimental-mount-plugins`, which contains comma-separated list of all volume plugins that should run their utilities in pods instead on the host. The list must include also all flex volume drivers. Without this option, controllers and kubelet would not know if a plugin should use a pod with mount utilites or directly host, especially on startup when the daemon set may not yet be fully deployed on all nodes.  * If so, it finds a pod in the dedicated namespace with label `mount.kubernetes.io/foo=true` (or `mount.kubernetes.io/flexvolume/foo=true` for flex volumes) and calls the volume plugin with `VolumeExec` pointing to the pod. All utilities that are executed by the volume plugin for attach/detach/provision/delete are executed in the pod as `kubectl exec <pod> xxx` (of course, we'll use clientset interface instead of executing `kubectl`).
```

Comment:
I admit I don't like this `experimental-mount-plugins` cmdline option. Can anyone find a bulletproof way, how kubelet / a controller can reliably find if a volume plugin should execute its utilities on the host or in a pod? Especially, when kubelet starts in a fresh cluster, the pod with mount utilities may not be running yet and kubelet *must* know if it should wait for it or not. Kubelet must not try to run the utilities on the host, because there may be wrong version or wrong configuration of the utilities.